internal/table/table.go

```diff
@@ -226,60 +282,36 @@ func (t *Table) Front(ctx context.Context, instanceID uint32) (*Job, error) {
 // disabled during execution, next_sched will advance but next_run will remain NULL.
 // Reversely, if doEnable is true but the job has a non-NULL next_run, the method call is a no-op.
-func (t *Table) UpdateNextRun(ctx context.Context, key Key, delay duration.Duration, randFactor float64, isRetry bool, doEnable bool) error {
-	if isRetry && doEnable {
-		return errors.New("isRetry and doEnable are mutually exclusive parameters")
-	}
-	tx, err := t.db.BeginTx(ctx, nil)
+func (t *Table) UpdateNextRun(ctx context.Context, key Key, randFactor float64, retryParams *retry.Retry) error {
+	tx, j, now, err := t.getForUpdate(ctx, key.Path, key.Body)
 	if err != nil {
 		return err
 	}
 	defer tx.Rollback() // nolint: errcheck
-	var locationName string
-	var now, nextSched time.Time
-	var nextRun sql.NullTime
-	s := "SELECT location, next_run, next_sched, IFNULL(CAST(? as DATETIME), UTC_TIMESTAMP())" +
-		" FROM " + t.name + " WHERE path = ? AND body = ?"
-	row := tx.QueryRowContext(ctx, s, t.Clk.NowUTC(), key.Path, key.Body)
-	if err := row.Scan(&locationName, &nextRun, &nextSched, &now); err != nil {
-		return fmt.Errorf("failed to get last run of job: %w", err)
-	}
-	if nextRun.Valid && doEnable {
+	if !j.NextRun.Valid {
+		// Job is disabled.
 		return nil
 	}
-	if locationName == "" {
-		locationName = time.UTC.String() // Default to UTC in case it's omitted somehow in the database.
-	}
-	location, err := time.LoadLocation(locationName)
-	if err != nil {
-		return err
-	}
-	now = now.In(location)
-	nextSched = nextSched.In(location)
 	switch {
-	case !t.FixedIntervals && !doEnable:
-		nextSched = delay.Shift(now)
-		if randFactor > 0 {
-			diff := randomize(nextSched.Sub(now), randFactor)
-			nextSched = now.Add(diff)
+	case retryParams != nil:
+		j.NextRun.Time = retryParams.NextRun(j.NextSched, now)
+	case t.FixedIntervals:
+		for j.NextSched.Before(now) {
+			j.NextSched = j.Interval.Shift(j.NextSched)
 		}
-	case t.FixedIntervals && !isRetry:
-		for nextSched.Before(now) {
-			nextSched = delay.Shift(nextSched)
+		j.NextRun.Time = j.NextSched
+	case !t.FixedIntervals:
+		j.NextSched = j.Interval.Shift(now)
+		if randFactor > 0 {
+			diff := randomize(j.NextSched.Sub(now), randFactor)
+			j.NextSched = now.Add(diff)
 		}
+		j.NextRun.Time = j.NextSched
 	}
-	switch {
-	case !nextRun.Valid && !doEnable:
-	case t.FixedIntervals && isRetry:
-		nextRun.Time = delay.Shift(now).UTC()
-	default:
-		nextRun.Time = nextSched.UTC()
-		nextRun.Valid = true
-	}
-	s = "UPDATE " + t.name + " " + // nolint: gosec
-		"SET next_run=?, next_sched=?, instance_id=NULL, location=? " +
+	s := "UPDATE " + t.name + " " + // nolint: gosec
+		"SET next_run=?, next_sched=?, instance_id=NULL " +
 		"WHERE path = ? AND body = ?"
-	_, err = tx.ExecContext(ctx, s, nextRun, nextSched.UTC(), locationName, key.Path, key.Body)
+	_, err = tx.ExecContext(ctx, s, j.NextRun.Time.UTC(), j.NextSched.UTC(), key.Path, key.Body)
```

Comment:
Maybe:

1. Have `nextRun()` take a `sql.NullTime` and if it's invalid, return it unchanged.
2. Separate out the setting of `j.NextRun`, so after `NextSched` is adjusted, then `if j.NextRun.Valid { j.NextRun.Time = j.NextSched` sort of thing.
3. On the Exec here, pass `j.NextRun` instead of `j.NextRun.Time` so that it can be null.

I think those tweaks would solve it.